stack/stack.mbt

```diff
@@ -0,0 +1,428 @@
+  @assertion.assert_eq(s.pop(), Some(3))?
+  @assertion.assert_eq(s.elements, Cons(2, Cons(1, Nil)))?
+  @assertion.assert_eq(s.len, 2)?
+  @assertion.assert_eq(s1.pop(), None)?
+  @assertion.assert_eq(s1.elements, Nil)?
+  @assertion.assert_eq(s1.len, 0)?
+}
+
+/// Drop the element at the top of the stack.
+/// Like pop, but does not return elements and does nothing if the Stack is empty.
+///
+/// # Example
+///
+/// ```
+/// let s: Stack[Int] = Stack::from_array([1,2,3])
+/// s.drop()
+/// println(s) // output: { elements: Cons(2, Cons(3, Nil)), len: 2 }
+/// ```
+pub fn drop[T](self : Stack[T]) -> Unit {
+  match self.elements {
+    Cons(_hd, tl) => {
+      self.elements = tl
+      self.len = self.len - 1
+    }
+    Nil => ()
+  }
+}
+
+test "drop" {
+  let s : Stack[Int] = Stack::[1, 2, 3]
+  s.drop()
+  @assertion.assert_eq(s.elements, Cons(2, Cons(1, Nil)))?
+  @assertion.assert_eq(s.len, 2)?
+}
+
+/// Only the top element of the stack is returned and will not be pop or drop.
+/// If there are elements in the stack, return `Some (the top element of the stack)`, otherwise return `None`.
+///
+/// # Example
+///
+/// ```
+/// let s: Stack[Int] = Stack::from_array([1,2,3])
+/// println(s.top()) // output: Some(3)
+/// println(s) // output: { elements: Cons(3, Cons(2, Cons(1, Nil))), len: 3 }
+/// ```
+pub fn top[T](self : Stack[T]) -> Option[T] {
+  match self.elements {
+    Cons(hd, _) => Some(hd)
+    Nil => None
+  }
+}
+
+test "top" {
+  let s : Stack[Int] = Stack::[1, 2, 3]
+  @assertion.assert_eq(s.top(), Some(3))?
+  @assertion.assert_eq(s.elements, Cons(3, Cons(2, Cons(1, Nil))))?
+  @assertion.assert_eq(s.len, 3)?
+}
+
+/// If stack is empty, return true, otherwise return false.
+///
+/// # Example
+///
+/// ```
+/// println(Stack::from_array([1, 2, 3]).is_empty()) // output: false
+/// println(Stack::new().is_empty()) // output: true
+/// ```
+pub fn is_empty[T](self : Stack[T]) -> Bool {
+  self.len == 0
+}
+
+test "is_empty" {
+  let empty : Stack[Unit] = Stack::new()
+  @assertion.assert_false(Stack::[1, 2, 3].is_empty())?
+  @assertion.assert_true(empty.is_empty())?
+}
+
+/// Returns the number of elements of the Stack
+pub fn length[T](self : Stack[T]) -> Int {
+  self.len
+}
+
+// Compares whether the elements in two lists are the same
+// pub fn equal[T](self: Stack[T], other: Stack[T]) -> Bool {
+//   if (self.len == other.len) {
+//     self.elements == other.elements
+//   } else {
+//     false
+//   }
+// }
+// 
+// pub fn op_equal[T](self: Stack[T], other: Stack[T]) -> Bool {
+//   self.equal(other)
+// }
+// 
+// test "equal" {
+//   let s1 : Stack[Int] = Stack::from_array([1, 2, 3])
+//   let s2 : Stack[Int] = Stack::from_array([1, 2, 3])
+//   @assertion.assert_eq(s1, s2)?
+// }
+
+/// Iterates over the stack.
+/// 
+/// # Example
+/// 
+/// ```
+/// Stack::from_array([1, 2, 3, 4, 5]).iter(print) // output: 54321
+/// ```
+pub fn iter[T](self : Stack[T], f : (T) -> Unit) -> Unit {
```

Comment:
I think the methods of the stack shouldn't manipulate the elements except for the top element. If someone needs it, they actually need a `List`.

Reply:
I think what you said makes sense. I thought so too at first, but I was just not sure whether I should add it, so I added it.

Reply:
I removed these functions in a new commit [569c3eb](https://github.com/moonbitlang/core/pull/40/commits/569c3eb8110044e78d8ad1d87b8ca8392b7efaa5)

Reply:
> I removed these functions in a new commit [569c3eb](https://github.com/moonbitlang/core/pull/40/commits/569c3eb8110044e78d8ad1d87b8ca8392b7efaa5)

Nice! Can you solve the conflicts before merging?